### PR TITLE
[fix] #3059: Simplify kura lock file code

### LINKED ISSUE #3059: Kura graceful shutdown
sometimes when executing benchmarks and sending a shutdown signal the following message appears:
```
2023-01-11T15:00:14.233857Z  WARN iroha_core::kura: Couldn't remove lockfile at Locked("/var/folders/hv/g1qbnmwj2ylc6nx9mdsbb2dr0000gn/T/.tmp6XYfkx"). Either the lockfile was deleted externally while Iroha instance was running, or there was an IO error. Check the block store with kura_inspector and then delete the lockfile manually. err=Tried to unlock block store by deleting lockfile at /var/folders/hv/g1qbnmwj2ylc6nx9mdsbb2dr0000gn/T/.tmp6XYfkx, but it couldn't be found
```

### PR DESCRIPTION
**Pull request by SamHSmith**
### Linked issue

Closes #3059  

### Checklist

- [ ] I've read `CONTRIBUTING.md`
- [ ] I've used the standard signed-off commit format (or will squash just before merging)
- [ ] All applicable CI checks pass (or I promised to make them pass later)
- [ ] (optional) I've written unit tests for the code changes
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: core/src/kura.rs [hunk 1]
```diff
@@ -62,7 +62,7 @@ impl Kura {
         block_store_path: &Path,
         debug_output_new_blocks: bool,
     ) -> Result<Arc<Self>> {
-        let mut block_store = BlockStore::new(block_store_path).lock()?;
+        let mut block_store = BlockStore::new(block_store_path, false);
```
Comment (appetrosyan) [inline core/src/kura.rs:65]: Consider 
```suggestion
        let mut block_store = BlockStore::new(block_store_path, BlockStoreType::Unlocked);
```

File: core/src/kura.rs [hunk 2]
```diff
@@ -84,10 +84,9 @@ impl Kura {
     /// Create a kura instance that doesn't write to disk. Instead it serves as a handler
     /// for in-memory blocks only.
     pub fn blank_kura_for_testing() -> Arc<Kura> {
-        let block_store = BlockStore::fake_locked_for_tests(PathBuf::new());
         Arc::new(Self {
             mode: Mode::Strict,
-            block_store: Mutex::new(block_store),
+            block_store: Mutex::new(BlockStore::new(&PathBuf::new(), true)),
```
Comment (appetrosyan) [inline core/src/kura.rs:89]:
```suggestion
            block_store: Mutex::new(BlockStore::new(&PathBuf::new(), BlockStoreType::Locked)),
```

File: core/src/kura.rs [hunk 3]
```diff
@@ -536,9 +415,48 @@ pub struct BlockIndex {
...
+                        }
+                    }
+                    _ => Err(Error::IO(e, path)),
+                }
+                .expect("Kura must be able to lock the blockstore");
+            }
+        }
+        BlockStore {
+            path_to_blockchain: path.to_path_buf(),
+        }
```
Comment (appetrosyan) [inline core/src/kura.rs:457]:
```suggestion
     else {
        BlockStore {
            path_to_blockchain: path.to_path_buf(),
        }
     }
```

The control flow is a bit confusing with this level of nesting.

File: core/src/kura.rs [hunk 4]
```diff
@@ -536,9 +415,48 @@ pub struct BlockIndex {
...
+            let path = path.join(LOCK_FILE_NAME);
+            if let Err(e) = fs::File::options()
+                .read(true)
+                .write(true)
+                .create_new(true)
+                .open(path.clone())
+            {
+                match e.kind() {
+                    std::io::ErrorKind::AlreadyExists => Err(Error::Locked(path)),
+                    std::io::ErrorKind::NotFound => {
```
Comment (appetrosyan) [inline core/src/kura.rs:431]: One of the worst designed features of the standard library.

**Review by appetrosyan — COMMENTED** [review]
Like it. Less code. In principle, could be done with a `const generic` and I'd honestly prefer an `enum` to make things a bit easier to understand, but no major objections.

File: core/src/kura.rs (hunk 3 above)
Comment (SamHSmith) [reply core/src/kura.rs:461]: I reordered to make it slightly easier on the eye.

File: core/src/kura.rs [hunk 5]
```diff
@@ -449,81 +391,17 @@ pub struct BlockCount {
...
-    /// - IO errors
-    /// - lockfile already exists
-    pub fn lock(self) -> Result<BlockStore<Locked>> {
-        Ok(BlockStore {
-            path_to_blockchain: self.path_to_blockchain.lock()?,
-        })
+impl Drop for BlockStore {
+    fn drop(&mut self) {
+        let path = self.path_to_blockchain.join(LOCK_FILE_NAME);
+        let _ = fs::remove_file(path); // we don't care if this succeeds or not
```
Comment (0x009922) [inline core/src/kura.rs:404]: Shouldn't it at least print a warning/debug message in case of failure?

File: core/src/kura.rs [hunk 6]
```diff
@@ -536,9 +414,53 @@ pub struct BlockIndex {
...
+        if matches!(already_locked, LockStatus::Unlocked) {
+            let path = path.join(LOCK_FILE_NAME);
+            if let Err(e) = fs::File::options()
+                .read(true)
+                .write(true)
+                .create_new(true)
+                .open(path.clone())
+            {
+                match e.kind() {
+                    _ => Err(Error::IO(e, path)),
```
Comment (0x009922) [inline core/src/kura.rs:434]: I am probable blind, but match arms evaluation depends on order ([example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=10a0a792c96012f37cde12e27528ee01)). I wonder why rustc doesn't print a warning about unreachable arms here.

File: core/src/kura.rs [hunk 7]
```diff
@@ -536,9 +414,53 @@ pub struct BlockIndex {
...
-/// Operations available both on locked and unlocked store,
-/// i.e. only read operations.
-impl<L: Lock> BlockStore<L> {
+pub enum LockStatus {
+    Locked,
+    Unlocked,
+}
+
+impl BlockStore {
+    /// Create a new read-only block store in `path`.
```
Comment (0x009922) [inline core/src/kura.rs:423]: Should this docstring contain information about possible panic inside?

File: core/src/kura.rs (hunk 6 above)
Comment (pesterev) [reply core/src/kura.rs:434]: > I wonder why rustc doesn't print a warning about unreachable arms here.

it does i've checked

File: core/src/kura.rs [hunk 8]
```diff
@@ -536,9 +414,53 @@ pub struct BlockIndex {
...
+                                    .open(path.clone())
+                                {
+                                    Err(Error::IO(e, path))
+                                } else {
+                                    Ok(())
+                                }
+                            }
+                        }
+                    }
+                }
```
Comment (pesterev) [inline core/src/kura.rs:455]:
```suggestion
                match e.kind() {
                    std::io::ErrorKind::AlreadyExists => Err(Error::Locked(path)),
                    std::io::ErrorKind::NotFound => std::fs::create_dir_all(&path)
                        .map_err(|e| Error::MkDir(e, path.clone()))
                        .and_then(|_| {
                            fs::File::options()
                                .read(true)
                                .write(true)
                                .create_new(true)
                                .open(&path)
                                .map_err(|e| Error::IO(e, path))
                        }),
                    _ => Err(Error::IO(e, path)),
                }
```
it seems matching have to be fixed, also you can simplify it using combinators and remove redundant clones.

File: core/src/kura.rs (hunk 5 above)
Comment (SamHSmith) [reply core/src/kura.rs:404]: No. The purpose of this PR is to not do that.

Comment (0x009922) [reply core/src/kura.rs:404]: I don't really understand why it shouldn't be done, but I am fine with that.

**Review by 0x009922 — APPROVED** [review]

**Review by pesterev — APPROVED** [review]

**Pull request merged**